File: Leetcode100/leetcode100_python/TrappingRainWater4.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        result = 0
        n = len(height)
        max_left, max_right = 0, 0
        for i in range(n):
            max_left = max(max_left, height[i])
            max_right = max(max_right, height[n - i - 1])
            result += max_left + max_right - height[i]
        return result - max_left * n


if __name__ == "__main__":
    height = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.trap(height))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/trapping_rain_water/TrappingRainWater4.java
package main.java.com.jiang.leetcode100.trapping_rain_water;

import java.util.Scanner;

class Solution4 {
    public int trap(int[] height) {
        int result = 0;
        int max_left = 0, max_right = 0;
        int left = 0, right = height.length - 1;
        while(left < right) {
            if(height[left] < height[right]) {
                max_left = Math.max(max_left, height[left]);
                result += max_left - height[left];
                left++;
            } else {
                max_right = Math.max(max_right, height[right]);
                result += max_right - height[right];
                right--;
            }
        }
        return result;
    }
}

public class TrappingRainWater4 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] height = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            height[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution4().trap(height));
    }
}